src/SFML/Window/WindowImpl.cpp

```diff
@@ -175,42 +176,59 @@ void WindowImpl::setMaximumSize(const std::optional<Vector2u>& maximumSize)
 
 
 ////////////////////////////////////////////////////////////
-Event WindowImpl::popEvent(bool block)
+Event WindowImpl::popEventImpl()
 {
+    Event event;
+
+    if (!m_events.empty())
+    {
+        event = m_events.front();
+        m_events.pop();
+    }
+
+    return event;
+}
+
+
+////////////////////////////////////////////////////////////
+Event WindowImpl::popEventBlocking(Time timeout)
+{
+    const auto startTime = std::chrono::steady_clock::now();
+
+    const auto timedOut = [&]
+    { return timeout != Time::Zero && (std::chrono::steady_clock::now() - startTime) < timeout.toDuration(); };
+
     // If the event queue is empty, let's first check if new events are available from the OS
     if (m_events.empty())
     {
         // Get events from the system
-        processJoystickEvents();
-        processSensorEvents();
-        processEvents();
+        populateEventQueue();
+    }
 
-        // In blocking mode, we must process events until one is triggered
-        if (block)
-        {
-            // Here we use a manual wait loop instead of the optimized
-            // wait-event provided by the OS, so that we don't skip joystick
-            // events (which require polling)
-            while (m_events.empty())
-            {
-                sleep(milliseconds(10));
-                processJoystickEvents();
-                processSensorEvents();
-                processEvents();
-            }
-        }
+    // Here we use a manual wait loop instead of the optimized
+    // wait-event provided by the OS, so that we don't skip joystick
+    // events (which require polling)
+    while (m_events.empty() && !timedOut())
+    {
+        sleep(milliseconds(10));
```

Comment:
This sleep duration imposes a sort of minimum timeout. If you want a timeout of 1 millisecond like in the tests you're still going to wait > 10 ms. Perhaps we document that timeouts under 10 milliseconds won't really work as intended. Alternatively maybe we make this duration some fraction of the provided timeout?

Reply:
Choosing timeouts in the range of milliseconds defeats the purpose of the timeout. While they technically could set it, I don't see it as problematic to have a not documented 10ms delay

Reply:
@eXpl0it3r: how about making the poll interval a parameter as well, and mentioning the relationship between timeout and polling in the documentation?

Reply:
No, this is an implementation detail. If we didn't decide to track sf::Joystick states via event data, this wouldn't even exist and we'd be using OS-level waiting.